First naïve per-session caching for JPA map store

ahus1 opened the pull request:
Closes #14938

Cache hits and misses can be checked via the following conditional breakpoints in 

Results: 

* Login/Logout on Account console: no hits (two ignored due to dirty session)
* Listing clients in admin UI: 13 misses, 39 hits
* Listing client scopes in admin UI: 19 misses, 5 hits
* Listing users in admin UI: 28 misses, 16 hits
* Listing realm roles in admin UI: 13 misses, 2 hits 

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

ahus1 committed:
First naïve per-session caching for JPA map store

Closes #14938

ahus1 commented:
/rerun

ahus1 commented:
/rerun

ahus1 commented:
/rerun

hmlnarik commented on model/map/src/main/java/org/keycloak/models/map/storage/QueryParameters.java:
```diff
@@ -69,6 +70,21 @@ public QueryParameters<M> orderBy(SearchableModelField<M> searchableModelField,
+
+    @Override
+    public int hashCode() {
+        // there is currently no equals method for the ModelCriteriaBuilder, take its String representation as a substitute.
```
There is no guarantee that `toString` would be defined on the respective `ModelCriteriaBuilder`.
Please update JPA `ModelCriteriaBuilder` to include proper equals / hashcode instead.

ahus1 replied:
Hi @hmlnarik - I pushed the next commit that doesn't rely on that comparison any more. Instead it first maps it to a JPA criteria query and then extracts the caching key from there. 

You mention that `ModelCriteriaBuilder` doesn't need to define a `toString`. At the same time `QueryParameters` always uses a `DefaultModelCriteria` which does have a `toString()`. Should the `QueryParameters` constructor be changed to accept a different class?

hmlnarik replied:
Thank you for the changes, please consider this comment resolved from my POV.

> Should the QueryParameters constructor be changed to accept a different class?

Not now, but it is possible in the future.

If `JpaModelCriteriaBuilder` defined `equals` and `hashCode`, perhaps we could leverage those instead of converting criteria to a `String`. On the other hand I see it as a potential next-step optimization, and would like to see the code in the current shape to be merged, as I believe it should already have visible impact.

ahus1 replied:
I gave it a thought to make `JpaModelCriteriaBuilder` implement `equals` and `hashCode`, but then abandoned it when I didn't find a simple way to make `JpaPredicateFunction` (one of the children `JpaModelCriteriaBuilder`) implement them for all of its variants. The corresponding element in `DefaultModelCriteria`, in the `ModelCriteriaNode`, implements `toString` which I used in the previous attempt.

ahus1 replied:
@hmlnarik - with #15038, the impact of the caching is now smaller and stated at be start of this issue; yet there is still some impact on every call on the admin UI.

hmlnarik replied:
I believe there is also impact of this PR on areas not explicitly affected by #15038 like all kinds of sessions which I suspect significant enough for incorporating this PR. Eventually would be nice to see this suspicion confirmed by a benchmark run.

ahus1 committed:
Don't rely on DefaultModeLCriteria in equals/hashCode

Instead, map this to JPA query and then create the cache lookup key from there.

Closes #14938

ahus1 commented:
/rerun

ahus1 commented:
/rerun

pedroigor commented:
/rerun

hmlnarik commented:
@ahus1 Is this ready to be switched from draft to a regular PR?

ahus1 commented:
@hmlnarik - now as the build is green it is now ready for review.

hmlnarik reviewed (approved):
Thank you for the PR!
